pytest/tests/sanity/upgradable.py

```diff
@@ -21,87 +19,65 @@
 import utils
 
 ONE_NEAR = 1000000000000000000000000
+NUM_VALIDATORS = 4
+EPOCH_LENGTH = 40
+NODE_PREFIX = "test"
 
-_EXECUTABLES = None
 
-TIMEOUT = 10
-
-
-def get_executables() -> branches.ABExecutables:
-    global _EXECUTABLES
-    if _EXECUTABLES is None:
-        _EXECUTABLES = branches.prepare_ab_test()
-        logger.info(f"Latest mainnet release is {_EXECUTABLES.release}")
-    return _EXECUTABLES
-
-
-def get_proto_version(exe: pathlib.Path) -> (int, int):
+def get_proto_version(exe: pathlib.Path) -> int:
     line = subprocess.check_output((exe, '--version'), text=True)
-    m = re.search(r'\(release (.*?)\) .* \(protocol ([0-9]+)\)', line)
+    m = re.search(r'.* \(protocol ([0-9]+)\)', line)
     assert m, (f'Unable to extract protocol version number from {exe};\n'
                f'Got {line.rstrip()} on standard output')
-    return m.group(1), int(m.group(2))
+    logger.info(f'Protocol version {m.group(1)} found in {exe}')
+    return int(m.group(1))
 
 
 class TrafficGenerator(threading.Thread):
     """ A thread which keeps sending transactions to random addresses until stopped. """
 
     def __init__(self, rpc_node: cluster.LocalNode, **kwargs) -> None:
         super().__init__(**kwargs)
-        self._lock = threading.Lock()
+        random.seed(2025)
         self._rpc_node = rpc_node
         self._stopped = False
-        self._paused = False
-        self._failed_txs = 0
+        self._failed = False
         self._acc1 = f"000000.{self._rpc_node.signer_key.account_id}"
         self._acc2 = f"zzzzzz.{self._rpc_node.signer_key.account_id}"
+        self._nonce = 1_000_000
 
     def run(self) -> None:
         logger.info("Starting traffic generator")
-        random.seed(2025)
         while not self._stopped:
-            if self._paused:
-                time.sleep(1)
-                continue
-            try:
-                with self._lock:
-                    self.send_transfer()
-                    self.call_test_contracts()
-            except Exception:
-                traceback.print_exc()
-                self._failed_txs += 1
+            self._with_retry(self.send_transfer)
+            self._with_retry(self.call_test_contracts)
         logger.info("Traffic generator stopped")
 
+    def join(self, timeout):
+        assert not self._failed
+        return super().join(timeout=timeout)
+
+    def _with_retry(self, fn):
+        for i in range(5):
+            try:
+                return fn()
+            except Exception as e:
+                logger.error(f"Failed txn try {i}", exc_info=True)
+        self._failed = True
```

Comment:
I think it would be more robust if you re-raise the exception from the last try. It's all too easy to forget about checking __failed. It's a minor thing, don't mind it if it's a lot of work or nasty code.